aten/src/ATen/native/cpu/HistogramKernel.cpp

```diff
@@ -0,0 +1,179 @@
+                /* When bin_edges is known to be a linear progression, maps elt to
+                 * the appropriate bin via simple division.
+                 */
+                pos = static_cast<int64_t>((elt - leftmost_bin_edge)
+                        / (rightmost_bin_edge - leftmost_bin_edge)
+                        * (numel_be - 1));
+
+                /* Ensures consistency with bin_edges by checking the bins to the left and right
+                 * of the selected position. Necessary for cases in which an element very close
+                 * to a bin edge may be misclassified by simple division.
+                 */
+                if (algorithm == LINEAR_INTERPOLATION_WITH_LOCAL_SEARCH) {
+                    int64_t pos_min = std::max(static_cast<int64_t>(0), pos - 1);
+                    int64_t pos_max = std::min(pos + 2, numel_be);
+                    pos = std::upper_bound(data_be + pos_min, data_be + pos_max, elt) - data_be - 1;
+                }
+            } else {
+                TORCH_INTERNAL_ASSERT(false);
+            }
+
+            // Unlike other bins, the rightmost bin includes its right boundary
+            if (pos == (numel_be - 1)) {
+                pos -= 1;
+            }
+
+            // In the unweighted case, the default weight is 1
+            input_t wt = accessor_wt.has_value() ? accessor_wt.value()[i] : static_cast<input_t>(1);
+            data_out_local[pos] += wt;
+        }
+
+        // Locks and updates the common output
+        const std::lock_guard<std::mutex> lock(data_out_mutex);
+        for (int64_t i = 0; i < numel_be - 1; i++) {
+            data_out[i] += data_out_local[i];
+        }
+    });
+}
+
+/* Some pre- and post- processing steps for the main algorithm.
+ * Initializes hist to 0, calls into the main algorithm, and normalizes output if necessary.
+ */
+template<BIN_SELECTION_ALGORITHM bin_algorithm>
+void histogram_out_cpu_template(const Tensor& self, const c10::optional<Tensor>& weight, bool density,
+        Tensor& hist, const Tensor& bin_edges) {
+    hist.fill_(0);
+
+    const int64_t numel_in = self.numel();
+    const Tensor reshaped_input = self.reshape({numel_in});
+
+    const auto reshaped_weight = weight.has_value()
+            ? c10::optional<Tensor>(weight.value().reshape({numel_in}))
+            : c10::optional<Tensor>();
+
+    switch (self.scalar_type()) {
+        case ScalarType::Double: {
+            histogram_cpu_contiguous<double, bin_algorithm>(
+                  hist, bin_edges.contiguous(), reshaped_input, reshaped_weight);
+            break;
+        }
+        case ScalarType::Float: {
+            histogram_cpu_contiguous<float, bin_algorithm>(
+                  hist, bin_edges.contiguous(), reshaped_input, reshaped_weight);
+            break;
+        }
+        default:
+            TORCH_CHECK(false, "torch.histogram: not supported on CPU for dtype ", self.dtype());
+    }
+
+    if (density) {
```

Comment:
Add a comment here -- maybe the next comment should just be pulled up to explain what this conditional does?